src/tree/index.tsx

```diff
@@ -0,0 +1,400 @@
+	initialExpanded?(id: string): void;
+	initialChecked?(id: string): void;
+	onSelect?(id: string): void;
+	onCheck?(id: string, checked: boolean): void;
+	onExpand?(id: string, expanded: boolean): void;
+}
+
+export interface TreeChildren {
+	(node: TreeNodeOption): RenderResult;
+}
+
+interface TreeCache {
+	activeNode?: string;
+	selectedNode?: string;
+	expandedNodes: string[];
+	checkedNodes: string[];
+	hasBeenExpanded: boolean;
+	hasBeenChecked: boolean;
+}
+
+const resource = createResourceMiddleware<TreeNodeOption>();
+const icache = createICacheMiddleware<TreeCache>();
+const factory = create({ theme, icache, diffProperty, focus, resource })
+	.properties<TreeProperties>()
+	.children<TreeChildren | undefined>();
+
+export default factory(function({
+	id,
+	middleware: { theme, icache, diffProperty, focus, resource },
+	properties,
+	children
+}) {
+	diffProperty(
+		'selectedNode',
+		properties,
+		({ selectedNode: current }, { selectedNode: next }) => {
+			if ((current || next) && current !== next) {
+				icache.set('selectedNode', next);
+			}
+		}
+	);
+	diffProperty(
+		'expandedNodes',
+		properties,
+		({ expandedNodes: current }, { expandedNodes: next }) => {
+			if ((current || next) && current !== next) {
+				icache.set('expandedNodes', next || []);
+			}
+		}
+	);
+	diffProperty(
+		'checkedNodes',
+		properties,
+		({ checkedNodes: current }, { checkedNodes: next }) => {
+			if ((current || next) && current !== next) {
+				icache.set('checkedNodes', next || []);
+			}
+		}
+	);
+
+	const { getOrRead, createOptions, isLoading, meta } = resource;
+	const {
+		checkable = false,
+		selectable = false,
+		onSelect,
+		onCheck,
+		onExpand,
+		initialChecked,
+		initialExpanded,
+		disabledNodes,
+		resource: { template },
+		parentSelection = false
+	} = properties();
+	const classes = theme.classes(css);
+	const defaultRenderer = (n: TreeNodeOption) => n.value;
+	const [itemRenderer] = children();
+
+	const activeNode = icache.get('activeNode');
+	const selectedNode = icache.get('selectedNode');
+	const expandedNodes = icache.getOrSet('expandedNodes', []);
+	const checkedNodes = icache.getOrSet('checkedNodes', []);
+	const hasBeenExpanded = icache.getOrSet('hasBeenExpanded', false);
+	const hasBeenChecked = icache.getOrSet('hasBeenChecked', false);
+	const shouldFocus = focus.shouldFocus();
+
+	function activateNode(id: string) {
+		icache.set('activeNode', id);
+	}
+
+	function selectNode(id: string) {
+		icache.set('selectedNode', id);
+		onSelect && onSelect(id);
+	}
+
+	function checkNode(n: string, checked: boolean) {
+		if (checked) {
+			checkedNodes.push(n);
+		} else {
+			checkedNodes.splice(checkedNodes.indexOf(n), 1);
+		}
+		if (checked) {
+			hasBeenChecked === false && initialChecked && initialChecked(n);
+			icache.set('hasBeenChecked', true);
+		}
+		icache.set('checkedNodes', checkedNodes);
+		onCheck && onCheck(n, checked);
+	}
+
+	function expandNode(id: string) {
+		expandedNodes.push(id);
```

Comment:
it's better not to save references to icache values to be unsed inside callbacks etc as they can become stale and result in errors.
In the case of changing the `expandedNodes` cache we should be doing something like the below in order to ensure the latest values are used. Please ensure the same is done elsewhere.

```tsx
icache.set('expandedNodes', (currentExpanded) => [ ...currentExpanded, id ]);
```

Reply:
I think we need to be careful here with auto-expanding when the user has supplied an `expandedNodes` controlling property. If this is present then we should not be updating the cache here and visually expanding the node, instead should be simply calling out to the `onExpand` callback and reacting to the new property change as with all other properties when the parent updates the `expandedNodes` property. I think this may be the case with other controlled properties within this PR.